wapitiCore/attack/mod_https_redirect.py

```diff
@@ -0,0 +1,91 @@
+from typing import Optional
+from urllib.parse import urlparse, urlunparse
+import socket
+
+from httpx import RequestError
+
+from wapitiCore.attack.attack import Attack
+from wapitiCore.definitions.https_redirect import NAME, WSTG_CODE
+from wapitiCore.definitions.internal_error import WSTG_CODE as INTERNAL_ERROR_WSTG_CODE
+from wapitiCore.language.vulnerability import Messages
+from wapitiCore.net import Request
+from wapitiCore.net.response import Response
+from wapitiCore.main.log import log_red, log_orange
+
+def test_port(address: str, dest_port: int, timeout: float = None) -> bool:
+    """Check if dest_port is open on address"""
+    try:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+            sock.settimeout(timeout)
+            if sock.connect_ex((address, dest_port)) == 0:
+                return True
+            return False
+    except (OSError, ValueError):
+        return False
+
+# This module check whether HTTP requests are redirected to HTTPS or not
+class ModuleHttpsRedirect(Attack):
+    """Check for HTTPS redirections."""
+    name = "https_redirect"
+    MSG_VULN_NO_REDIRECT = "No HTTPS redirection"
+    MSG_VULN_REDIRECT = "Redirected to HTTP"
+
+    async def must_attack(self, request: Request, response: Optional[Response] = None):
+        if self.finished:
+            return False
+
+        url = urlparse(request.url)
+        if not test_port(url.hostname, url.port or 80, self.crawler.timeout):
```

Comment:
There is something to fix about the tested port, here in `must_attack` or/and in `attack`.

If the URL received as parameter is for example http://perdu.com:8888/ then the port being tested would be 8888 and it seems ok but does it makes sense as Wapiti tested the URL while crawling ?

However if the URL is http**s**://perdu.com:8843/ then the tested port would be 8843 which is incorrect (it should be 80) while it makes more sense to test the port.

Also do we want to test all URLs found by the crawler (http or https) or just consider https urls ? It makes more sense to consider only https URLs as someone scanning an http-only website on purpose certainly already knows the risks while someone feeding Wapiti with an https URL might be interested in downgrade attacks.

Wdyt ?

Reply:
Here is the new behavior : 

- if http url provided, it will stop the module after the first request and if there is no redirection to https, it will also raise a vuln
- if specific port (8080 or 8443 for instance) is provided, it will stop the module after the first request, if it is https we will probably get a 400 error, if it is http it will raise a vuln only if there is no redirection to https
- in other cases, it will test all the requests